### ISSUE TITLE
Regression Gamma Loss Gradient calculation is incorrect with weight

### ISSUE THREAD

**Issue by katrat123** [microsoft/LightGBM#4174]
## Description
In src/objective/regression_objective.hpp, line 698 the weight term is only being applied to a single term, rather than the entire expression.  As a result, the weighted calculation of gamma objective is incorrect : 
gradients[i] = static_cast<score_t>(1.0 - label_[i] / std::exp(score[i]) * weights_[i]);

## Reproducible example
Lightgbm simplified code (incorrect) : (1.0 - y / exp(p) * w)
Xgboost simplified code (correct) : (1 - y / exp(p)) * w
Lightgbm simplifed code (correct) : (1-y/exp(p))*w
Tweedie, rho = 2 (gamma) simplified code : ((-y * exp(-1 * p) + exp(0 * p)) * w === (1 – y / exp(p)) * w

**Comment by jameslamb** [comment]
Thanks for opening an issue! Are you interested in contributing a fix?

**Comment by shiyu1994** [comment]
@katrat123 Yes, that's a mistake. Thanks for reporting that!

**Comment by katrat123** [comment]
Sure can do.  Just create a branch and submit a pull request?

Kate Ely
Analyst II, Data Science – Pricing Sophistication
Liberty Mutual Insurance
Safeco Plaza 1001 4th Ave, Seattle, WA 98154
Direct Dial: 425-519-5632

***@***.***D73151.276743F0]

From: James Lamb ***@***.***>
Sent: Tuesday, April 13, 2021 5:11 PM
To: microsoft/LightGBM ***@***.***>
Cc: Ely, Kathryn <Kathryn.Ely@LibertyMutual.com>; Author ***@***.***>
Subject: {EXTERNAL} Re: [microsoft/LightGBM] Regression Gamma Loss Gradient calculation is incorrect with weight (#4174)


Thanks for opening an issue! Are you interested in contributing a fix?

**Comment by jameslamb** [comment]
Yes please! The process is:

1. Fork this repository
2. create a branch on your fork
3. push changes to thay branch
4. open a pull request at https://github.com/microsoft/LightGBM/pulls. End the title of the pull request with "(fixes #4174)"

Let me know if you run into any issues! Thanks again for helping to improve LightGBM 🙌

**Comment by StrikerRUS** [comment]
Gently ping @katrat123

**Comment by katrat123** [comment]
OOO next week; will do upon my return.

Thanks!
Kate

Kate Ely
Analyst II, Data Science – Pricing Sophistication
Liberty Mutual Insurance
Safeco Plaza 1001 4th Ave, Seattle, WA 98154
Direct Dial: 425-519-5632

***@***.***D73DB8.9D4635E0]

From: Nikita Titov ***@***.***>
Sent: Wednesday, April 28, 2021 5:36 AM
To: microsoft/LightGBM ***@***.***>
Cc: Ely, Kathryn <Kathryn.Ely@LibertyMutual.com>; Mention ***@***.***>
Subject: {EXTERNAL} Re: [microsoft/LightGBM] Regression Gamma Loss Gradient calculation is incorrect with weight (#4174)


Gently ping @katrat123<https://github.com/katrat123>

**Comment by mayer79** [comment]
@StrikerRUS : Weighted Gamma regression is a quite typical model in insurance claim modelling. The response is the claim cost divided by the number of claims, while the weight is the number of claims. I would love to see this fixed. 
@katrat123 : If you do not have started and are short of time, I can do that pull request!

**Comment by katrat123** [comment]
Have been OOO, would be great if you can complete the pull request.

Thanks!
Kate

Kate Ely
Analyst II, Data Science – Pricing Sophistication
Liberty Mutual Insurance
Safeco Plaza 1001 4th Ave, Seattle, WA 98154
Direct Dial: 425-519-5632

***@***.***D747D8.CE9A3680]

From: Michael Mayer ***@***.***>
Sent: Thursday, May 13, 2021 12:37 AM
To: microsoft/LightGBM ***@***.***>
Cc: Ely, Kathryn <Kathryn.Ely@LibertyMutual.com>; Mention ***@***.***>
Subject: {EXTERNAL} Re: [microsoft/LightGBM] Regression Gamma Loss Gradient calculation is incorrect with weight (#4174)


@StrikerRUS<https://github.com/StrikerRUS> : Weighted Gamma regression is a quite typical model in insurance claim modelling. The response is the claim cost divided by the number of claims, while the weight is the number of claims. I would love to see this fixed.
@katrat123<https://github.com/katrat123> : If you do not have started and are short of time, I can do that pull request!
